dubbo-remoting/dubbo-remoting-api/src/main/java/org/apache/dubbo/remoting/api/Connection.java

```diff
@@ -149,14 +155,16 @@ public void onGoaway(Channel channel) {
 
     public void onConnected(Channel channel) {
         this.channel.set(channel);
+        // This indicates that the connection is available.
+        this.connectedFuture.complete(CONNECTED_OBJECT);
         channel.attr(CONNECTION).set(this);
         if (logger.isInfoEnabled()) {
             logger.info(String.format("%s connected ", this));
         }
     }
 
-    public void connectSync() {
-        this.initPromise.awaitUninterruptibly(this.connectTimeout);
+    public void connectSync() throws InterruptedException, ExecutionException, TimeoutException {
+        this.connectedFuture.get(this.connectTimeout, TimeUnit.MILLISECONDS);
```

Comment:
Why change the origin modify. Is there some problem when use spin to check isAvailabel()?
If we use future to realize sync mechanism, the connectedFuture is stateful, if the connect close and then reconnect, the connectFuture can't gurantee sync mechanism.

Reply:
Using CompletableFuture code looks more elegant. And I also think that connectedFuture should be reset when the channel is unavailable. I modify the code.